### PR TITLE
add partial index support (pg only)

### PR DESCRIPTION
**Pull request by lure**
**Limitation:**
- postgresql only is supported 
- `currentDialect.existingIndices` does not return the full filterCondition, just a placeholder to mark the index as a partial - the text to expression parser is not added. 

**How to Create :**
```kotlin
    object Test : Table("partialIndex") {
        val id = integer("id")
        val name = varchar("name", 128)
        init {
            index(columns = arrayOf(id, name)){
                (id greaterEq  3) and ( name.isNotNull() )
            }
        }
    }

```

**How to drop**
```kotlin
// Variant 1
// implies the table has just a single index
val index = currentDialect.existingIndices(partialIndexTable)[partialIndexTable].first()

// variant 2, if constraint 
val index = Index(columns = listOf(partialIndexTable.value, partialIndexTable.name), unique = false)
// variant 2, not constraint.
val index = Index(columns = listOf(partialIndexTable.value, partialIndexTable.name), unique = false){Op.TRUE} 

// and finally 
exec { index.dropStatement() } 
``` 

---

**Details user must be aware of**
There are three cases of indexed data in PG that we are interested in: Index, Unique Index, Unique Constraint. 
To create [unique] index we execute `CREATE [UNIQUE] INDEX `
To create Constraint, this form is used: `ALTER TABLE ...  ADD CONSTRAINT ...         UNIQUE (team, "member")`
The unique constraint always creates the unique index. 
This knowledge is important to understand the following invariants: 
- Unique constraints and indexes (remember, constraint always creates the index under the hood) can not specify index type, it is always B-tree. 
- Unique constraint can not be partial 
- Unique index can be partial. 

as a flow 
```
 if isUniq and notPartial -> Constraint
 If isUniq -> Unique Index, withType is ignored 
 if withType -> Index, isUniq ignored
 else -> Index
``` 

**Something to consider**

Unique Constraint is not mandatory to create referencing foreign key, contrary to a common disbelieve. However, according to PG doc, to is a constraint is a more idiomatic way. 

```sql
db=# create table master(id integer not null, name text ); 
CREATE TABLE
db=# create table detail (did integer not null, details text, mid integer references master(id)); 
ERROR:  there is no unique constraint matching given keys for referenced table "master"
db=# \d master 
               Table "public.master"
 Column |  Type   | Collation | Nullable | Default 
--------+---------+-----------+----------+---------
 id     | integer |           | not null | 
 name   | text    |           |          | 

db=# create unique index master_id_unique on master (id);
CREATE INDEX
db=# \d master
               Table "public.master"
 Column |  Type   | Collation | Nullable | Default 
--------+---------+-----------+----------+---------
 id     | integer |           | not null | 
 name   | text    |           |          | 
Indexes:
    "master_id_unique" UNIQUE, btree (id)

db=# create table detail (did integer not null, details text, mid integer references master(id));
CREATE TABLE
crcltdevdb=# \d detail 
               Table "public.detail"
 Column  |  Type   | Collation | Nullable | Default 
---------+---------+-----------+----------+---------
 did     | integer |           | not null | 
 details | text    |           |          | 
 mid     | integer |           |          | 
Foreign-key constraints:
    "detail_mid_fkey" FOREIGN KEY (mid) REFERENCES master(id)
``` 

Since Exposed resorted to the usage of unique constraint since it's inception, we can not completely switch to the Unique Indexes usage only. Additional reason to keep things slightly more complicated is that it will not break the backward compatibility with exposed.

### CODE REVIEW

**Comment by lure** [comment]
CC:  @turansky  @e5l  @joc-a

**Comment by e5l** [comment]
@joc-a could you take a look?

File: exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/shared/ddl/CreateIndexTests.kt [hunk 1]
```diff
@@ -83,4 +82,68 @@ class CreateIndexTests : DatabaseTestsBase() {
...
+
+            fun List<Column<*>>.names(): Set<String> { return map { identity(it) }.toSet() }
+            fun getIndexes(): List<Index> {
+                db.dialect.resetCaches()
+                return currentDialect.existingIndices(partialIndexTable)[partialIndexTable].orEmpty()
+            }
+
+            val dropIndex = Index(columns = listOf(partialIndexTable.value, partialIndexTable.name), unique = false).dropStatement().first()
+            kotlin.test.assertTrue(dropIndex.startsWith("DROP INDEX "), "Unique partial index must be created and dropped as index")
+            val dropUniqConstraint = Index(columns = listOf(partialIndexTable.anotherValue), unique = true).dropStatement().first()
```
Comment (joc-a) [inline exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/shared/ddl/CreateIndexTests.kt:140]: nit (for consistency across the codebase):
```suggestion
            val dropUniqueConstraint = Index(columns = listOf(partialIndexTable.anotherValue), unique = true).dropStatement().first()
```

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/vendors/Default.kt [hunk 2]
```diff
@@ -1013,30 +1015,61 @@ abstract class VendorDialect(
...
+                    .append(" WHERE ").append(it)
+                    .toString()
+            }
+        } else {
+            null
+        }
+    }
+
+    /**
+     * The uniquieness might be required for foreign constraints
```
Comment (joc-a) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/vendors/Default.kt:1080]: nit:
```suggestion
     * Uniqueness might be required for foreign key constraints.
```

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/vendors/Default.kt [hunk 3]
```diff
@@ -1061,30 +1064,61 @@ abstract class VendorDialect(
...
+     *
+     * In postgresq (https://www.postgresql.org/docs/current/indexes-unique.html) Uniq means btree only.
```
Comment (joc-a) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/vendors/Default.kt:1082]: nit:
```suggestion
     * In PostgreSQL (https://www.postgresql.org/docs/current/indexes-unique.html), UNIQUE means B-tree only.
```

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/vendors/Default.kt [hunk 4]
```diff
@@ -1061,30 +1064,61 @@ abstract class VendorDialect(
...
+
+        val maybeFilterCondition = filterCondition(index) ?: ""
+
         return when {
-            index.unique -> {
+            // uniq and no filter -> constraint, the type is not supported
+            index.unique && maybeFilterCondition.isEmpty() -> {
                 "ALTER TABLE $quotedTableName ADD CONSTRAINT $quotedIndexName UNIQUE $columnsList"
             }
+            // uniq and filter -> index only, the type is not supported
```
Comment (joc-a) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/vendors/Default.kt:1099]: nit (for consistency across the codebase):
```suggestion
            // unique and filter -> index only, the type is not supported
```

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/vendors/Default.kt [hunk 5]
```diff
@@ -1061,30 +1064,61 @@ abstract class VendorDialect(
...
+            index.unique -> {
+                "CREATE UNIQUE INDEX $quotedIndexName ON $quotedTableName $columnsList$maybeFilterCondition"
+            }
+            // type -> can't be uniq or constraint
```
Comment (joc-a) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/vendors/Default.kt:1103]: nit (for consistency across the codebase):
```suggestion
            // type -> can't be unique or constraint
```

File: exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/shared/ddl/CreateIndexTests.kt [hunk 6]
```diff
@@ -83,4 +82,68 @@ class CreateIndexTests : DatabaseTestsBase() {
...
+                    flag eq true
+                }
+                index(columns = arrayOf(value, name)) {
+                    (name eq "aaa") and (value greaterEq 6)
+                }
+                uniqueIndex(columns = arrayOf(anotherValue))
+            }
+        }
+
+        withDb(TestDB.POSTGRESQL) {
```
[the remark this answers is not in the document]
Comment (lure) [reply exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/shared/ddl/CreateIndexTests.kt:106]: Honestly, no idea.

[the remark this answers is not in the document]
Comment (joc-a) [reply exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/shared/ddl/CreateIndexTests.kt:106]: I think we should, because even though `PostgreSQLNGDialect` extends `PostgreSQLDialect`, in TestDB, they are separate.

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/vendors/Mysql.kt [hunk 7]
```diff
@@ -278,7 +278,8 @@ open class MysqlDialect : VendorDialect(dialectName, MysqlDataTypeProvider, Mysq
         }
     }
 
-    override fun dropIndex(tableName: String, indexName: String): String = "ALTER TABLE $tableName DROP INDEX $indexName"
+    override fun dropIndex(tableName: String, indexName: String, isUnique: Boolean, isPartial: Boolean): String =
+        "ALTER TABLE ${identifierManager.quoteIfNecessary(tableName)} DROP INDEX ${identifierManager.quoteIfNecessary(indexName)}"
```
[the remark this answers is not in the document]
Comment (lure) [reply exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/vendors/Mysql.kt:282]: The existing approach failed for some corner cases of camel case naming.

File: exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/shared/ddl/CreateIndexTests.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (joc-a) [reply exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/shared/ddl/CreateIndexTests.kt]: @lure The line you added [here](https://github.com/JetBrains/Exposed/pull/1748/files#:~:text=return%20if%20(currentDialect%20is%20PostgreSQLDialect)%20%7B) means that the changes you introduced will apply to both `PostgreSQLDialect` _and_ `PostgreSQLNGDialect`, since `PostgreSQLNGDialect` extends `PostgreSQLDialect`, which means tests need to cover both. Thanks for adding it in the test 👍🏼

**Comment by lure** [comment]
any news? 
@joc-a   @bog-walk  @Tapac

**Review by joc-a — APPROVED** [review]

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/Table.kt [hunk 8]
```diff
@@ -962,22 +975,25 @@ open class Table(name: String = "") : ColumnSet(), DdlAware {
...
+    fun <T> Column<T>.uniqueIndex(customIndexName: String? = null): Column<T> =
+        index(customIndexName, true)
 
     /**
      * Creates a unique index.
      *
      * @param columns Columns that compose the index.
      */
-    fun uniqueIndex(vararg columns: Column<*>): Unit = index(null, true, *columns)
+    fun uniqueIndex(vararg columns: Column<*>, filterCondition: FilterCondition = null): Unit =
```
Comment (bog-walk) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/Table.kt:986]: Please add a new `@param filterCondition` as was done in the previous function.

**Review by bog-walk — CHANGES_REQUESTED** [review]
@lure Thank you for your work on moving this PR forward.

In addition to the 2 KDocs additions I've requested, could you please run `apiDump` as you did for your other PR?

**Review by bog-walk — APPROVED** [review]

**Comment by lure** [comment]
@bog-walk  @joc-a @e5l 
Sssoooooooo?

**Review by e5l — APPROVED** [review]
